### ISSUE TITLE
No such file or directory

### ISSUE THREAD

**Issue by col** [boldpoker/edeliver#57]
I'm getting the following errors when I try to do a build.

```
$ mix edeliver build release
deps/edeliver/bin/edeliver: line 26: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/defaults: No such file or directory
deps/edeliver/bin/edeliver: line 27: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/output: No such file or directory
deps/edeliver/bin/edeliver: line 28: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/deprecations: No such file or directory
deps/edeliver/bin/edeliver: line 29: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/core: No such file or directory
deps/edeliver/bin/edeliver: line 30: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/common: No such file or directory
deps/edeliver/bin/edeliver: line 31: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/generated: No such file or directory
deps/edeliver/bin/edeliver: line 32: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/erlang: No such file or directory
deps/edeliver/bin/edeliver: line 39: /Users/col/Projects/test_app/deps/edeliver
/Users/col/Projects/test_app/deps/edeliver/libexec/erlang-init: No such file or directory
deps/edeliver/bin/edeliver: line 41: __capture_runtime_configs: command not found
deps/edeliver/bin/edeliver: line 42: __find_all_strategies: command not found
deps/edeliver/bin/edeliver: line 45: __load_app_config: command not found
deps/edeliver/bin/edeliver: line 46: __default_app_config: command not found
deps/edeliver/bin/edeliver: line 47: __apply_runtime_configs: command not found
deps/edeliver/bin/edeliver: line 48: __load_strategy: command not found
deps/edeliver/bin/edeliver: line 49: __remote_hosts: command not found
deps/edeliver/bin/edeliver: line 50: __check_config: command not found
deps/edeliver/bin/edeliver: line 56: begin: command not found
deps/edeliver/bin/edeliver: line 57: run: command not found
deps/edeliver/bin/edeliver: line 58: finish: command not found
```

I've had a look into it and it seems to come down to this line in ``/bin/edeliver``

```
BASE_PATH="$(cd "$DIR/.." && pwd)"
```

Somehow this is actually setting BASE_PATH to: 

```
/Users/col/Projects/test_app/deps/edeliver /Users/col/Projects/test_app/deps/edeliver
```

I think this can be fixed by using this command instead.

```
BASE_PATH="$(cd "$DIR/.." && `pwd`)"
```

HTH

**Comment by bharendt** [comment]
Thanks for reporting and investigating this and providing a solution. 

If I understand right, the output of `$(cd "$DIR/.." && pwd)` returns the "`BASE_PATH`" twice in your case (separated by a newline).

This statement is used to find the path where edeliver is installed (usually `your-project/deps/edeliver`) by also respecting file-system links on several systems. This is done by changing into the parent directory of the script and printing that path using `pwd`.

If this results in duplicate output of the path which is set to the `BASE_PATH` env, that would mean, that the `cd "$DIR/.."` command also prints the directory to stdout. This is usually not the case. __Are you maybe using an alias for the `cd` command? Or which (unix-like) system are you using__?
Would be nice if you could figure out what could cause the `cd` command on your system to output the path, because this might break also other functionality of edeliver for you or at least the output. I would __suggest to change that__ line __to__:
```sh
BASE_PATH="$(cd "$DIR/.." >/dev/null 2>&1 && pwd)"
```
in your case. [Substituting](http://tldp.org/LDP/abs/html/commandsub.html) the `pwd` command (by using back-quotes) would just hide its output because it actually produces an error printed to stderr, because `pwd` (quoted/substituted) tries to execute a command with the name of the base path and would  produce output like that:

```
$ `pwd`
-bash: /Users/col/Projects/test_app/deps/edeliver: is a directory
```

This means, this would work only in your case (where the `cd` command prints the destination directory), but not for other users of edeliver.

**Comment by col** [comment]
Yeah I thought something like might be the case but I figured giving you my
fix would at least help to explain what's going on.

Yes I agree that something strange is afoot. I'm running zsh so maybe
something is affecting cd.

Thanks for the suggestions. I'll do some further investigation and let you
know.

Cheers

Col

**Comment by bharendt** [comment]
Hi @col: Did you find something out about the `cd` command prining the destination path in your case? 

Maybe typing `type -t cd` can help you. It shows e.g. whether it is an __alias or a function__ and `type cd` shows detailed information.

```sh
$ type cd
cd is a shell builtin
$ cd () { command cd "$@" && pwd; }
$ cd ~
/Users/username
$ type cd
cd is a function
cd () 
{ 
    command cd "$@" && pwd
}
$ unset -f cd
```

Or maybe you have __set the__ `CDPATH` __env__? This forces the `cd` command to output the "found" directory:

```sh
$ export CDPATH=~
$ cd ~
$ cd tmp/
/Users/username/tmp
```

In any case, I think it makes sense to hide the output of the `cd` command as suggested above:

```sh
BASE_PATH="$(cd "$DIR/.." >/dev/null 2>&1 && pwd)"
```

Would you like to add it to your pull request?

**Comment by bharendt** [comment]
The same applies to [bin/edeliver:11](https://github.com/boldpoker/edeliver/blob/master/bin/edeliver#L11)

**Comment by col** [comment]
Sorry for wasting your time on this. I did have CDPATH set. I've unset it now and everything is working fine now. 

I am curious now though. Seems like having CDPATH set wouldn't be that uncommon. Is there a way to get this to worth either way?

**Comment by bharendt** [comment]
Thanks.

**Comment by col** [comment]
Done.

**Comment by bharendt** [comment]
Published it as [hex version 1.1.6](https://hex.pm/packages/edeliver/1.1.6). Thanks for reporting this and locating the error.
